### DISCUSSION TITLE
How to get detection and tracking data from each frame of the stream (like box location, id, class, etc.)?

### DISCUSSION THREAD

**Discussion by SenchaBrest** [ultralytics/ultralytics#1653] [Q&A]
Hello. I have been working on my project which was based on yolov5. There, using yolov5, I detected objects on the stream and tracked them using the tracking algorithm. I recently decided to upgrade to a newer version, yolov8, which already has botsort and bytetrack. Due to the fact that there are already built-in algorithms for tracking, I will not need to embed extraneous ones.

However, I can't figure out how to do it. I need to know the information after detection and tracking from each frame and enter it into the database, but I can't find how to do it, because for me the function call scheme inside the repository looks very confusing.

Could you tell me how to get detection and tracking data from each frame of the stream (like box location, id, class, etc.)

**Comment by glenn-jocher** [comment]
@SenchaBrest see https://docs.ultralytics.com/modes/track for details.

<img width="1024" src="https://github.com/ultralytics/assets/raw/main/yolov8/banner-integrations.png">

Object tracking is a task that involves identifying the location and class of objects, then assigning a unique ID to
that detection in video streams.

The output of tracker is the same as detection with an added object ID.

## Available Trackers

The following tracking algorithms have been implemented and can be enabled by passing `tracker=tracker_type.yaml`

* [BoT-SORT](https://github.com/NirAharon/BoT-SORT) - `botsort.yaml`
* [ByteTrack](https://github.com/ifzhang/ByteTrack) - `bytetrack.yaml`

The default tracker is BoT-SORT.

## Tracking

Use a trained YOLOv8n/YOLOv8n-seg model to run tracker on video streams.

```python
from ultralytics import YOLO

# Load a model
model = YOLO('yolov8n.pt')  # load an official detection model
model = YOLO('yolov8n-seg.pt')  # load an official segmentation model
model = YOLO('path/to/best.pt')  # load a custom model

# Track with the model
results = model.track(source="https://youtu.be/Zgi9g1ksQHc", show=True) 
results = model.track(source="https://youtu.be/Zgi9g1ksQHc", show=True, tracker="bytetrack.yaml") 
```

As in the above usage, we support both the detection and segmentation models for tracking and the only thing you need to
do is loading the corresponding (detection or segmentation) model.

## Configuration

### Tracking

Tracking shares the configuration with predict, i.e `conf`, `iou`, `show`. More configurations please refer
to [predict page](https://docs.ultralytics.com/modes/predict/).
    
```python
from ultralytics import YOLO

model = YOLO('yolov8n.pt')
results = model.track(source="https://youtu.be/Zgi9g1ksQHc", conf=0.3, iou=0.5, show=True) 
```

### Tracker

We also support using a modified tracker config file, just copy a config file i.e `custom_tracker.yaml`
from [ultralytics/tracker/cfg](https://github.com/ultralytics/ultralytics/tree/main/ultralytics/tracker/cfg) and modify
any configurations(expect the `tracker_type`) you need to.

```python
from ultralytics import YOLO

model = YOLO('yolov8n.pt')
results = model.track(source="https://youtu.be/Zgi9g1ksQHc", tracker='custom_tracker.yaml') 
```

Please refer to [ultralytics/tracker/cfg](https://github.com/ultralytics/ultralytics/tree/main/ultralytics/tracker/cfg) page.

**Comment by dh031200** [comment]
@SenchaBrest

I don't know if this is the right way, but you can get information like this.

```python
model = YOLO("yolov8n.pt")
results = model.track(source=src, tracker="botsort.yaml", verbose=False, stream=True)
for r in results:
    for tracker in detect_model.predictor.trackers:
        print(tracker.__dict__) # tracker information
        print()
        for strack in tracker.tracked_stracks:
            print(strack.__dict__)  # Each obeject information
```

Output:
### Tracker information
```
{'tracked_stracks': [OT_1_(1-1), OT_2_(1-1), OT_3_(1-1), OT_4_(1-1), OT_5_(1-1)], 
 'lost_stracks': [],
 'removed_stracks': [],
 'frame_id': 1,
 'args': IterableSimpleNamespace(tracker_type='botsort', track_high_thresh=0.5, track_low_thresh=0.1, new_track_thresh=0.6, track_buffer=30, 
  match_thresh=0.8, cmc_method='sparseOptFlow', proximity_thresh=0.5, appearance_thresh=0.25, with_reid=False),
 'max_time_lost': 30,
 'kalman_filter': <ultralytics.tracker.utils.kalman_filter.KalmanFilterXYWH object at 0x7f59c844f370>,
 'proximity_thresh': 0.5,
 'appearance_thresh': 0.25,
 'gmc': <ultralytics.tracker.utils.gmc.GMC object at 0x7f59c844c550>}
```

### Obeject information
```
{'_tlwh': array([     1260.5,      583.32,      124.22,      267.34], dtype=float32), 'kalman_filter': <ultralytics.tracker.utils.kalman_filter.KalmanFilterXYWH object at 0x7f59c844f370>, 'mean': array([     1322.6,      716.99,      124.22,      267.34,           0,           0,           0,           0], dtype=float32), 'covariance': array([[     154.32,           0,           0,           0,           0,           0,           0,           0],
       [          0,      714.69,           0,           0,           0,           0,           0,           0],
       [          0,           0,      154.32,           0,           0,           0,           0,           0],
       [          0,           0,           0,      714.69,           0,           0,           0,           0],
       [          0,           0,           0,           0,       60.28,           0,           0,           0],
       [          0,           0,           0,           0,           0,      279.17,           0,           0],
       [          0,           0,           0,           0,           0,           0,       60.28,           0],
       [          0,           0,           0,           0,           0,           0,           0,      279.17]]), 'is_activated': True, 'score': 0.91442204, 'tracklet_len': 0, 'cls': 15.0, 'idx': 0.0, 'smooth_feat': None, 'curr_feat': None, 'features': deque([], maxlen=50), 'alpha': 0.9, 'track_id': 1, 'state': 1, 'frame_id': 1, 'start_frame': 1}
```

**Reply by glenn-jocher** [reply]
Thanks for sharing your approach, @dh031200.

@SenchaBrest, the `results` object returned by `model.track()` contains the tracking information you need for each frame. You can access attributes like `r.boxes.id`, `r.boxes.xyxy`, `r.boxes.cls`, and `r.boxes.conf` by iterating through the results list, as shown in the [Multi-Object Tracking documentation](https://docs.ultralytics.com/modes/track/).
